MATLAB/Algorithms/LSQR.m

```diff
@@ -0,0 +1,273 @@
+%--------------------------------------------------------------------------
+% This file is part of the TIGRE Toolbox
+% 
+% Copyright (c) 2015, University of Bath and 
+%                     CERN-European Organization for Nuclear Research
+%                     All rights reserved.
+%
+% License:            Open Source under BSD. 
+%                     See the full license at
+%                     https://github.com/CERN/TIGRE/blob/master/LICENSE
+%
+% Contact:            tigre.toolbox@gmail.com
+% Codes:              https://github.com/CERN/TIGRE/
+% Coded by:           Ander Biguri 
+%--------------------------------------------------------------------------
+
+%%
+
+% msl: I assume this x is either an array of 0 of the appropriate
+% dimensions or an initial guess
+
+[verbose,x,QualMeasOpts,gpuids,redundancy_weights]=parse_inputs(proj,geo,angles,varargin);
+
+% msl: no idea of what this is. Should I check?
+measurequality=~isempty(QualMeasOpts);
+qualMeasOut=zeros(length(QualMeasOpts),niter);
+if redundancy_weights
+    % Data redundancy weighting, W_r implemented using Wang weighting
+    % reference: https://iopscience.iop.org/article/10.1088/1361-6560/ac16bc
+    
+    num_frames = size(proj,3);
+    W_r = redundancy_weighting(geo);
+    W_r = repmat(W_r,[1,1,num_frames]);
+    % disp('Size of redundancy weighting matrix');
+    % disp(size(W_r));
+end
+
+% Paige and Saunders //doi.org/10.1145/355984.355989
+
+% Enumeration as given in the paper for 'Algorithm LSQR'
+% (1) Initialize 
+u=proj-Ax(x,geo,angles,'Siddon','gpuids',gpuids);
+normr = norm(u(:),2);
+u = u/normr;
+
+beta = normr;
+phibar = beta;
+
+% msl: I have not checked 'redundancy_weights', do they have to do with 
+% mismatched transpose? 
+if redundancy_weights 
+    v=Atb(W_r .* u,geo,angles,'matched','gpuids',gpuids);
+else
+    v=Atb(u,geo,angles,'matched','gpuids',gpuids);
+end
+
+alpha = norm(v(:),2); % msl: do we want to check if it is 0? 
+v = v/alpha;
+rhobar = alpha;
+w = v;
+
+normAtr = beta*alpha; % Do we want this? ||A^T r_k||
+% msl: do we want to check for convergence? In well posed problems it would
+% make sense, not sure now.
+
+errorL2=zeros(1,niter); % msl: is this error the residual norm ? 
+
+% (2) Start iterations 
+for ii=1:niter
+    x0 = x;
+    if (ii==1 && verbose);tic;end
+    
+    % (3)(a)
+    u = Ax(v,geo,angles,'Siddon','gpuids',gpuids) - alpha*u;
+    beta = norm(u(:),2);
+    u = u / beta;
+    
+    % (3)(b)
+    if redundancy_weights % msl: Is this right here?
+        v = Atb(W_r .* u,geo,angles,'matched','gpuids',gpuids) - beta*v;
+    else
+        v = Atb(u,geo,angles,'matched','gpuids',gpuids) - beta*v;
+    end
+    alpha = norm(v(:),2);
+    v = v / alpha;    
+
+    % (4)(a-g)
+    rho = sqrt(rhobar^2 + beta^2);
+    c = rhobar / rho;
+    s = beta / rho;
+    theta = s * alpha;
+    rhobar = - c * alpha; % msl: I am overwriting a matlab function, maybe change name

```

Comment:
Agree, any suggestion?

Reply:
actually that is for rho. Change to rho_var or something like that